Add fix for marketo renaming of g-captcha-response

pmahnke opened the pull request:
## Done

- Added js that creates a new field called grecaptcharesponse to hold the inforomation in g-recaptcha-response, because Marketo secretly doesn't allow hyphens

## QA

- Check out this feature branch
- Run the site using the command `./run serve`
- View the site locally in your web browser at: [http://0.0.0.0:8001/openstack](http://0.0.0.0:8001/openstack#get-in-touch) or any other marketo form
- Run through the following [QA steps](https://github.com/canonical-webteam/practices/blob/master/workflow/qa-steps.md)
- Fill in the form and check the recaptcha box and see that it gets to marketo and that the marketo webhook tests the submission


## Issue / Card

Fixes #4888

## Screenshot

![image](https://user-images.githubusercontent.com/441217/55193029-cd719880-519d-11e9-9345-4d0fa999b25a.png)

webteam-app commented:
Starting demo at: https://www-ubuntu-com-canonical-websites-pr-4889.run.demo.haus/

nottrobin commented on static/js/dynamic-contact-form.js:
```diff
@@ -200,6 +200,19 @@
+      function() {
+        var input = document.createElement('input');
+        input.type = 'hidden';
+        input.name = "grecaptcharesponse";
```
Do you think we should try something more like "LastReCAPTCHAUserResponse" just for consistency with the other solutions out there?

nottrobin commented on templates/shared/forms/_recaptcha.html:
```diff
@@ -0,0 +1,15 @@
+    function() {
+      var input = document.createElement('input');
+      input.type = 'hidden';
+      input.name = "grecaptcharesponse";
```
Do you think we should try something more like "lastReCAPTCHAUserResponse" just for consistency with the other solutions out there?

nottrobin replied:
Or would you rather not rock the boat, now that we've got it working?

pmahnke replied:
let's leave it for now.

nottrobin commented on static/js/dynamic-contact-form.js:
```diff
@@ -200,6 +200,19 @@
       return message;
     }
 
+    // move g-recaptcha-response to marketo field
```
Why have you duplicated this here? It seems like everywhere that you're linking to this JS file, you're then also calling `{% include "shared/forms/_recaptcha.html" %}` right after, which contains this same JS block. Unless I missed something?

pmahnke replied:
because this page is already a giant js function?  the include has script tags, etc...

nottrobin replied:
That doesn't make any sense to me. That block of code requires `grecaptcha` to be defined. This is added to the page by `<script src="https://www.google.com/recaptcha/api.js" async defer></script>`, which is only included in `templates/shared/forms/_recaptcha.html`. So you, quite rightfully, add a `<script>` block which uses `grecaptcha` right after that line, inside `templates/shared/forms/_recaptcha.html`.

So why do you also need it here? The block in the include should be sufficient.

nottrobin replied:
You include both in every case:

``` bash
 $ rg -A 2 dynamic-contact-form.min.js
templates/shared/forms/interactive/_general.html
257:<script src="{% versioned_static 'js/build/dynamic-contact-form.min.js' %}"></script>
258-{% include "shared/forms/_recaptcha.html" %}

templates/shared/forms/interactive/_support.html
227:<script src="{% versioned_static 'js/build/dynamic-contact-form.min.js' %}"></script>
228-{% include "shared/forms/_recaptcha.html" %}

templates/shared/forms/interactive/_bootstack.html
372:<script src="{% versioned_static 'js/build/dynamic-contact-form.min.js' %}"></script>
373-{% include "shared/forms/_recaptcha.html" %}

templates/shared/forms/interactive/_kubernetes.html
288:<script src="{% versioned_static 'js/build/dynamic-contact-form.min.js' %}"></script>
289-{% include "shared/forms/_recaptcha.html" %}

templates/shared/forms/interactive/_robotics.html
241:<script src="{% versioned_static 'js/build/dynamic-contact-form.min.js' %}"></script>
242-{% include "shared/forms/_recaptcha.html" %}

templates/shared/forms/interactive/_embedded.html
224:<script src="{% versioned_static 'js/build/dynamic-contact-form.min.js' %}"></script>
225-{% include "shared/forms/_recaptcha.html" %}

templates/shared/forms/interactive/_openstack.html
374:<script src="{% versioned_static 'js/build/dynamic-contact-form.min.js' %}"></script>
375-{% include "shared/forms/_recaptcha.html" %}

templates/shared/forms/interactive/_kubeflow.html
203:<script src="{% versioned_static 'js/build/dynamic-contact-form.min.js' %}"></script>
204-{% include "shared/forms/_recaptcha.html" %}
```

nottrobin commented on templates/download/server/index.html:
```diff
@@ -39,7 +39,6 @@ <h2>Ubuntu Server {{ latest_release_with_point }}</h2>
           <p class="p-card__content">
             <a class="p-button--positive is-wide" href="/download/server/thank-you?version={{ latest_release_with_point }}&amp;architecture=amd64" onclick="dataLayer.push({'event' : 'GAEvent', 'eventCategory' : 'Download', 'eventAction' : 'Server', 'eventLabel' : '{{ latest_release }}', 'eventValue' : undefined });">Download</a></p>
             <p><small>For other versions of Ubuntu including torrents, the network installer, a list of local mirrors, and past releases <a href="/download/alternative-downloads">see our alternative downloads</a>.</small></p>
-          </p>
```
If you have time, you could reformat this by putting the `</p>`s on their own lines - to avoid the confusion which presumably led to the extra `</p>` in the first place.

pmahnke replied:
only for you...

nottrobin commented:
Awesome! :trophy:. A couple of comment inline.

pmahnke commented:
@nottrobin 	could you re-review?  I would love to get this out today.

nottrobin commented:
Did you miss `templates/engage/events.html`?

pmahnke commented:
@nottrobin another round

nottrobin commented on templates/download/server/thank-you.html:
```diff
@@ -44,96 +44,7 @@ <h1 style="font-weight: 100;">Thank you for downloading Ubuntu Server</h1>
-   * Given a form element, extract its data and its "action" target URL
-   * and submit the form using Ajax
-   */
-  backgroundSubmit = function(formElement, submitCallback) {
```
OMG is background submit dead?! That's wonderful news!

pmahnke replied:
no, I just moved it

nottrobin commented on templates/shared/forms/_recaptcha.html:
```diff
@@ -0,0 +1,15 @@
+      input.type = 'hidden';
+      input.name = "grecaptcharesponse";
+      input.value = grecaptcha.getResponse();
+      this.appendChild(input);
```
In testing I realised that if you submit the form with errors, this gets run, `grecaptcharesponse` gets added, but then the form doesn't get submitted. If you then submit the form again, another `grecaptcharesponse` gets added, and an array of 2 values is sent to Marketo. I haven't checked, but I imagine Marketo won't handle that well.

So I suggest this:

``` javascript
  // move g-recaptcha-response to marketo field
  var marketoForm = document.querySelector('form#mktoForm');
  marketoForm.addEventListener(
    'submit',
    function() {
      var input = marketoForm.querySelector('[name=grecaptcharesponse]')
      if (! input) {
        input = document.createElement('input');
        input.type = 'hidden';
        input.name = "grecaptcharesponse";
        this.appendChild(input);
      }
      input.value = grecaptcha.getResponse();
    }
  )
```

pmahnke commented:
@nottrobin 	now?

nottrobin commented:
For me the `grecaptcharesponse` is not being sent with the dynamic modal contact-us forms (the "contact us" link in the footer), only with the static ones.

@anthonydillon maybe you could help me get this fixed and out the door?

anthonydillon commented:
@nottrobin wow a lot has changed on this branch. I'll take a look shortly.

nottrobin commented:
@anthonydillon I didn't mean tonight

pmahnke commented:
@nottrobin it worked for me... did you get to the last page of the form to hit the google box?

![image](https://user-images.githubusercontent.com/441217/55748696-375c2e80-5a37-11e9-9eed-7e4cfd38a5f6.png)

nottrobin commented:
@pmahnke meant to talk to you about this yesterday. Now it seems to successfully submit the `gcaptcharesponse` field for me on the homepage - although we should check all of the dynamic forms really - but it doesn't work in the back-end.

I think all the forms need wiring up in Marketo the same way that the cloud contact for is wired, so we can probably do an end-to-end test.

nottrobin commented:
@pmahnke linkchecker helpfully pointed out that some files you deleted caused errors:

```
URL        `/download/iot/kvm'
Name       `\n              KVM\n            '
Parent URL http://127.0.0.1:8000/community/governance, line 841, col 38
Real URL   http://127.0.0.1:8000/download/iot/kvm
Check time 0.714 seconds
Result     Error: 500 Internal Server Error
```

![image](https://user-images.githubusercontent.com/519935/55780475-e59bbe80-5a9f-11e9-815f-ccd50ea2e5d9.png)

pmahnke commented:
@nottrobin can you look again? I added an include for background downloads and it all 'appears' to work for me.

caldav commented:
Please hold on, there is a lot of form performance reporting that depends on having the form id in the "mktoForm_<marketo ID>" format. The value is captured by Tag manager, to register a submit form event.

caldav commented on templates/shared/forms/_recaptcha.html:
```diff
@@ -0,0 +1,18 @@
+<script src="https://www.google.com/recaptcha/api.js" async defer></script>
+<script>
+  // move g-recaptcha-response to marketo field
+  var marketoForm = document.querySelector('form#mktoForm');
```
Could we do `document.querySelector("form[id^=mktoForm]")` instead? So we can keep form ids in the `mktoForm_<marketo ID>` format?

nottrobin replied:
If we have multiple forms on the same page, we'll also need `querySelectorAll` and a loop.

anthonydillon replied:
Or use a class?

anthonydillon replied:
For example: `document.querySelectorAll(".mktoForm")`

nottrobin replied:
Sure, and if we're going to do that, why not change `mkto` to `marketo`? We're not writing Go here.

caldav replied:
@nottrobin please don't change it in the id without a good reason though. IMHO, the amount of fixes required (eg. transitioning simple string contains to RegEx to blend new and old data) in GA, Datastudio and many spreadsheets, for the sake of not writing Go, is not worth it.

nottrobin replied:
@caldav Yeah, I can see that. I was referring to @anthonydillon's suggestion that we create a class - in which case we're choosing what to name it. If we were choosing the IDs from scratch then I would say `marketo-form` would be a better scheme, but it definitely seems like a lot of hassle to change it now. So I'm not advocating that.

anthonydillon replied:
I believe @nottrobin is referring to the class name not the id.

pmahnke commented:
@caldav do you want to look again?

nottrobin commented on templates/download/server/provisioning.html:
```diff
@@ -167,20 +167,4 @@ <h2>Technical documentation</h2>
 {% endblock content %}
 
 {% block footer_extra %}
```
If `{% block footer_extra %}` is empty, it's probably cleaner to simply remove it, no?

pmahnke commented:
@nottrobin, @anthonydillon and @caldav I did another refactor, and am using just one submit listener. I think it is all working, but I didn't test every single thing yet, but would love you to look at the _form_validation.html file and see if that is going to work.

- It get's rid of jquery (from most things) and relies on html5 validation... which might need to be polyfilled?
- It requires the g-recaptcha
- It has french and greman engage forms
- It does a wonky hide/show for the dynamic forms to see the modal thank-you

My tests show that it does two forms on a page ok.  Downloads when the hidden field (new) is there.  Redirects when the hidden field is there.  Opens the thank you modal for dynamic forms.

anthonydillon commented on templates/shared/forms/interactive/_bootstack.html:
```diff
@@ -297,7 +297,7 @@ <h3 class="p-heading--five">What would be most helpful?</h3>
       <div class="row u-no-padding">
         <div class="col-6">
-          <form action="https://pages.ubuntu.com/index.php/leadCapture/save" method="post" id="mktoForm">
+          <form action="https://pages.ubuntu.com/index.php/leadCapture/save" method="post" id="mktoForm_{{ formid }}" class="marketo-form">
```
This form needs `formVid`

anthonydillon commented on templates/shared/forms/interactive/_embedded.html:
```diff
@@ -149,7 +149,7 @@ <h3 class="p-heading--five">What are your platform preferences?</h3>
       <div class="row u-no-padding">
         <div class="col-6">
-          <form action="https://pages.ubuntu.com/index.php/leadCapture/save" method="post" id="mktoForm">
+          <form action="https://pages.ubuntu.com/index.php/leadCapture/save" method="post" id="mktoForm_{{ formid }}" class="marketo-form">
```
This form needs `formVid`

anthonydillon commented on templates/shared/forms/interactive/_kubeflow.html:
```diff
@@ -128,7 +128,7 @@ <h3 class="p-heading--five">Have we missed something?</h3>
       <div class="row u-no-padding">
         <div class="col-6">
-          <form action="https://pages.ubuntu.com/index.php/leadCapture/save" method="post" id="mktoForm">
+          <form action="https://pages.ubuntu.com/index.php/leadCapture/save" method="post" id="mktoForm_{{ formid }}" class="marketo-form">
```
This form needs `formVid`

anthonydillon commented on templates/shared/forms/interactive/_kubernetes.html:
```diff
@@ -213,7 +213,7 @@ <h3 class="p-heading--five">Have we missed something?</h3>
       <div class="row u-no-padding">
         <div class="col-6">
-          <form action="https://pages.ubuntu.com/index.php/leadCapture/save" method="post" id="mktoForm">
+          <form action="https://pages.ubuntu.com/index.php/leadCapture/save" method="post" id="mktoForm_{{ formid }}" class="marketo-form">
```
This form needs `formVid`

anthonydillon commented on templates/shared/forms/interactive/_openstack.html:
```diff
@@ -299,7 +299,7 @@ <h3 class="p-heading--five">What would be most helpful?</h3>
       <div class="row u-no-padding">
         <div class="col-6">
-          <form action="https://pages.ubuntu.com/index.php/leadCapture/save" method="post" id="mktoForm">
+          <form action="https://pages.ubuntu.com/index.php/leadCapture/save" method="post" id="mktoForm_{{ formid }}" class="marketo-form">
```
This form needs `formVid`

anthonydillon commented on templates/shared/forms/interactive/_robotics.html:
```diff
@@ -166,7 +166,7 @@ <h3 class="p-heading--five">What are your platform preferences?</h3>
       <div class="row u-no-padding">
         <div class="col-6">
-          <form action="https://pages.ubuntu.com/index.php/leadCapture/save" method="post" id="mktoForm">
+          <form action="https://pages.ubuntu.com/index.php/leadCapture/save" method="post" id="mktoForm_{{ formid }}" class="marketo-form">
```
This form needs `formVid`

anthonydillon commented on templates/shared/forms/interactive/_support.html:
```diff
@@ -152,7 +152,7 @@ <h3 class="p-heading--five">Tell us about your project or interest</h3>
       <div class="row u-no-padding">
         <div class="col-6">
-          <form action="https://pages.ubuntu.com/index.php/leadCapture/save" method="post" id="mktoForm">
+          <form action="https://pages.ubuntu.com/index.php/leadCapture/save" method="post" id="mktoForm_{{ formid }}" class="marketo-form"r>
```
This form needs `formVid` and a stray `r` after the class

pmahnke commented on templates/engage/de/openstack-made-easy.html:
```diff
@@ -45,94 +45,7 @@ <h3 class="p-muted-heading u-align--center">Wir erstellen Clouds für die führe
-      <script src="{{ ASSET_SERVER_URL }}f97fa297-stateCountry.js"></script>
-
-      <!-- /MARKETO FORM -->
+      {% include "engage/shared/_fr_engage_form.html" with id="3266" returnURL="https://pages.ubuntu.com/openstack-made-easy-de-TY.html" cta="eBook herunterladen" %}
```
good catch

pmahnke commented on templates/engage/shared/_de_engage_form.html:
```diff
@@ -0,0 +1,64 @@
+<!-- MARKETO FORM -->
+<form action="https://pages.ubuntu.com/index.php/leadCapture/save" method="post" id="mktoForm_{{ id }}" style="margin-top:-1em;">
```
I don't think I use it anymore... perhaps I should remove them everywhere

pmahnke commented on templates/shared/forms/_form-validation.html:
```diff
@@ -0,0 +1,175 @@
+
+    // recaptcha test
+    var recaptchaListElement = marketoForm.getElementsByClassName('g-recaptcha')[0];
+    var recaptchaWidgetId = recaptchaListElement.dataset.widgetId;
```
can you explain? before I set the variable to use the first element, or before I use it in the next line?

anthonydillon replied:
Just check that you found the element before trying to access is dataset. Otherwise it will error

pmahnke commented on templates/shared/forms/_form-validation.html:
```diff
@@ -0,0 +1,175 @@
+    }
+  }
+
+  backgroundSubmit = function(marketoForm, submitCallback) {
```
that broke it... so I left it.

caldav commented:
I was not able to get a lead in Marketo out of:
* the whitepaper form at https://www-ubuntu-com-canonical-websites-pr-4889.run.demo.haus/engage/finserv-multi-cloud?utm_source=takeover&utm_campaign=CY19_DC_Finance_eBook_451
* Maas form on server thank you

pmahnke commented:
@caldav I must have deleted the ```<input type="hidden" aria-hidden="true" aria-label="hidden field" name="munchkinId" class="mktoField" value="066-EOV-335" />```

doh.

can you try again?

webteam-app commented:
Starting demo at: https://www-ubuntu-com-canonical-web-and-design-pr-4889.run.demo.haus/

caldav commented:
Still nothing coming from https://www-ubuntu-com-canonical-websites-pr-4889.run.demo.haus/engage/finserv-multi-cloud

pmahnke commented:
@caldav I *think* the issue is you are testing the old demo service... we have moved it as we moved the orgs - canonical-web-and-design not canonical-websites

caldav commented:
@pmahnke good catch, switched demos and seeing some progress:
Out of 11 correct form fills, 5 landed:
Failures were at:
- Newsletter subscription form on https://www-ubuntu-com-canonical-web-and-design-pr-4889.run.demo.haus/download/desktop/thank-you (never got the success notification at the top of the page, no lead)
- MAAS ebook form on https://www-ubuntu-com-canonical-web-and-design-pr-4889.run.demo.haus/download/server/thank-you (download was triggered, no lead)

![Screenshot from 2019-04-29 10-06-35](https://user-images.githubusercontent.com/18480003/56883286-8578b680-6a66-11e9-8ae5-e99d8e56f8e1.png)

pmahnke commented:
@caldav I think it is working now... with a better thank you message for background submits

caldav commented:
@pmahnke it works :) :+1:

anthonydillon commented on templates/download/desktop/thank-you.html:
```diff
@@ -341,14 +346,14 @@ <h3 class="p-card--title">Select topics you&rsquo;re interested in</h3>
+      ) {
+        matchingItem = items[itemIndex];
       }
+    }
```
Traditionally you put this coding bracket on the same line as the sad face below

pmahnke replied:
ok, not my code

anthonydillon commented on templates/download/desktop/thank-you.html:
```diff
@@ -23,7 +23,7 @@
 {% endif %}
 
-{% if request.GET.newsletter == 'true' %}
+{% if request.GET.newsletter == ' true' %}
```
Small nik but could you remove the space before the true please

anthonydillon commented on templates/download/desktop/thank-you.html:
```diff
@@ -442,24 +447,24 @@ <h3 class="p-card--title">Select topics you&rsquo;re interested in</h3>
   document.querySelector('#contributions-form').addEventListener(
-    'submit',
-    sendContributionFormAnalytics
+  'submit',
```
Did you use prettier for this branch? There is some odd white space that has been removed.

pmahnke replied:
no, perhaps atom's auto indent lines

anthonydillon commented on templates/shared/forms/_form-validation.html:
```diff
@@ -0,0 +1,175 @@
+<!-- background submit download inlcude start -->
+<style>
+  .recaptcha-is-error {
+    border: 1px solid red;
```
Could you use the Vanilla colour here?

pmahnke replied:
moved to styles.css

anthonydillon commented on templates/shared/forms/_form-validation.html:
```diff
@@ -0,0 +1,175 @@
+    console.log("download_asset_url: " + download_asset_url);
+    console.log("return_url: " + return_url);
+    console.log("modal: " + isModal);
+    console.log("thankYouMessage: " + thankYouMessage);
```
Can you remove these console logs please

pmahnke replied:
done

anthonydillon commented on templates/shared/forms/_form-validation.html:
```diff
@@ -0,0 +1,175 @@
+
+    // add a thank-you notification to the page
+    // if someone submitted a form without a thank you action
+    var thankYouMessage = thankYouMessage;
```
You don't need to do this. That variable is already set.

anthonydillon commented on templates/shared/forms/_form-validation.html:
```diff
@@ -0,0 +1,175 @@
+    // if someone submitted a form without a thank you action
+    var thankYouMessage = thankYouMessage;
+    if (return_url === null || isModal === false) {
+      console.log("thank you: " + thankYouMessage);
```
Please remove console logs

pmahnke replied:
done

anthonydillon commented on templates/shared/forms/_form-validation.html:
```diff
@@ -0,0 +1,175 @@
+    let recaptchas = document.querySelectorAll("div[class^=g-recaptcha]");
+    recaptchas.forEach(function(field){
+      recaptchaWidgetId = grecaptcha.render(field, {'sitekey' : '6LfYBloUAAAAAINm0KzbEv6TP0boLsTEzpdrB8if'});
+      field.setAttribute("data-widget-id", recaptchaWidgetId);
```
You are using data-widgetId above. Should they be the same?

pmahnke replied:
no?
